torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,188 @@
+        r._tensor = tensor  # type: ignore[attr-defined]
+        return r
+
+    def __repr__(self):
+        return f"AsyncCollectiveTensor({self._tensor})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e: Any):
+            if isinstance(e, AsyncCollectiveTensor):
+                return wait_tensor(e._tensor)
+            return e
+
+        # TODO what happens when func is an inplace? (add_) -
+        # currently, seems like the mutation works on the underlying tensor,
+        # but it remains an AsyncCollectiveTensor so subsequent ops still 'unwrap' again.
+        # would be nice to fix this.
+
+        unwrapped_args = tree_map(unwrap, args)
+        unwrapped_kwargs = tree_map(unwrap, kwargs)
+
+        out = func(*unwrapped_args, **unwrapped_kwargs)
+        return out
+
+
+# TODO assert if ranks has duplicated entries
+def _all_reduce_cpu(self, reduceOp, tag, ranks, stride):
+    reduceOp = reduceOp.lower()
+    assert reduceOp == "sum", "Unable to convert str to ReduceOp, so only default sum works"
+    assert tag == "", "No support for non-empty comms tag"
+    assert len(ranks) % stride == 0, f"Ranks length ({len(ranks)}) must be divisible by stride ({stride})"
+
+    my_rank = dist.get_rank()
+    rank_set = None
+
+    for i in range(0, len(ranks), stride):
+        rank_set = ranks[i : i + stride]
+        if my_rank in rank_set:
+            my_ranks = rank_set
+
+    assert my_ranks is not None, "Called all_reduce with a set of ranks that doesn't include the current node"
+    assert my_rank in my_ranks, "Called all_reduce with a set of ranks that doesn't include the current node"
+
+    my_ranks.sort()
+
+    group = c10d._find_pg_by_ranks_and_tag(tag, my_ranks)
+    assert group is not None
+
+    inplace_tensor = self.clone()
+    work = dist.all_reduce(inplace_tensor, op=dist.ReduceOp.SUM, group=group, async_op=True)
+
+    global tensor_to_work
+    tensor_to_work.add(inplace_tensor, work)
+    return inplace_tensor
+
+c10_lib = torch.library.Library("aten", "IMPL",  "CPU")
+c10_lib.impl("all_reduce", _all_reduce_cpu)
+
+RANK_TYPES = Union[List[int], List[List[int]], dist.ProcessGroup, dt.DeviceMesh]
+
+def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str = ""):
+    """
+    Reduces the tensor data across all machines in such a way that all get
+    the final result.
+
+    The input tensor is left unmodified.
+
+    rank can be one of:
+        List[int]: ranks participating in the collective.
+        List[List[int]]: 2D mesh of ranks taking part of this collective in MPMD.
+        ProcessGroup: if you don't like all the new cool stuff or has some legacy weigthing on your shoulder.
+        DeviceMesh: Do a SPMD collective over all ranks of a collective
+        (DeviceMesh, int): Do a MPMD collective over one dimension of the DeviceMesh
+    """
+    if isinstance(group, list):
```

Comment:
lets make a helper for validating/normalizing a rankset, to return the list+stride?

Reply:
👍